source/server/http/admin.cc

```diff
@@ -529,15 +545,77 @@ Http::Code AdminImpl::handlerClusters(absl::string_view url, Http::HeaderMap& re
-                                        Buffer::Instance& response, AdminStream&) const {
-  envoy::admin::v2alpha::ConfigDump dump;
+void AdminImpl::addAllConfigToDump(envoy::admin::v2alpha::ConfigDump& dump,
+                                   const absl::optional<std::string>& mask) const {
+  for (const auto& key_callback_pair : config_tracker_.getCallbacksMap()) {
+    ProtobufTypes::MessagePtr message = key_callback_pair.second();
+    RELEASE_ASSERT(message, "");
+
+    if (mask.has_value()) {
+      Protobuf::FieldMask field_mask;
+      ProtobufUtil::FieldMaskUtil::FromString(mask.value(), &field_mask);
+      ProtobufUtil::FieldMaskUtil::TrimMessage(field_mask, message.get());
+    }
+
+    auto* config = dump.add_configs();
+    config->PackFrom(*message);
+  }
+}
+
+void AdminImpl::addResourceToDump(envoy::admin::v2alpha::ConfigDump& dump,
+                                  const absl::optional<std::string>& mask,
+                                  const std::string& resource) const {
   for (const auto& key_callback_pair : config_tracker_.getCallbacksMap()) {
     ProtobufTypes::MessagePtr message = key_callback_pair.second();
     RELEASE_ASSERT(message, "");
-    auto& any_message = *(dump.add_configs());
-    any_message.PackFrom(*message);
+
+    auto field_descriptor = message->GetDescriptor()->FindFieldByName(resource);
+    const Protobuf::Reflection* reflection = message->GetReflection();
+    if (!field_descriptor) {
+      continue;
+    } else if (!field_descriptor->is_repeated()) {
+      throw AdminException{
+          Http::Code::BadRequest,
+          fmt::format("{} is not a repeated field. Use ?mask={} to get only this field",
+                      field_descriptor->name(), field_descriptor->name())};
+    }
+
+    auto repeated = reflection->GetRepeatedPtrField<Protobuf::Message>(*message, field_descriptor);
+    for (Protobuf::Message& msg : repeated) {
+      if (mask.has_value()) {
+        Protobuf::FieldMask field_mask;
+        ProtobufUtil::FieldMaskUtil::FromString(mask.value(), &field_mask);
+        ProtobufUtil::FieldMaskUtil::TrimMessage(field_mask, &msg);
+      }
+      auto* config = dump.add_configs();
+      config->PackFrom(msg);
+    }
+
+    // We found the desired resource so there is no need to continue iterating over
+    // the other keys.
+    return;
+  }
+
+  throw AdminException{Http::Code::NotFound, fmt::format("{} not found in config dump", resource)};
+}
+
+Http::Code AdminImpl::handlerConfigDump(absl::string_view url, Http::HeaderMap& response_headers,
+                                        Buffer::Instance& response, AdminStream&) const {
+  Http::Utility::QueryParams query_params = Http::Utility::parseQueryString(url);
+  const auto resource = resourceParam(query_params);
+  const auto mask = maskParam(query_params);
+
+  envoy::admin::v2alpha::ConfigDump dump;
+
+  if (resource.has_value()) {
+    try {
```

Comment:
Not a huge fan of adding more exceptions on the data plane; this adds an additional case to reason about when analyzing Envoy's exception safety.

Reply:
I was wondering if this would be controversial. I'm not familiar with idiomatic c++ error handling practices but I've switched to returning an optional instead of throwing an exception.